### PR TITLE
AGPUSH-971 : Advanced Analytics, Server Side 

### PR DESCRIPTION
**Pull request by sebastienblanc**
New PR Based on the closed https://github.com/aerogear/aerogear-unifiedpush-server/pull/523

* rebase afer JMS landed on master branch :white_check_mark: 
* `request.getHeader("push-identifier");` header needs a different prefix (aerogear-push-id) :white_check_mark: 
* move the `extractPushIdentifier()` to the `HttpRequestUtil` util :white_check_mark: 
* first/last opened will stay as is, right ? :white_check_mark: 
* `message.getMessage().getUserData().put("push-identifier",pushMessageInformation.getId());` needs to be done at outer/different/custom level - not on the envelope/payload (e.g. the user data)  :white_check_mark:  (has been removed from "our" user-data but for each push network I can only end up in their respective "custom" section)

* leverage the MetricsCollector instead of new VariantMetricInformationDao on this PR : *Will be handled  when Registration will use messaging*
* push identifier: docment it on the endpoint doc : *REST Endpoint has been updated but update to the endpoint doc will be in a aerogear.org PR*
* JavaDoc on `NotificationSenderCallback` : *not relevant anymore*

* `pushMessageInformation.setAppOpenCounter` needs atomicLong or AtomicRef : *still open for discussion, we have now the "simple" version (incrementing a long)*

@lfryc @edewit @lfryc can you take a look ?

### CODE REVIEW

**Comment by matzew** [comment]
> leverage the MetricsCollector instead of new VariantMetricInformationDao on this PR : Will be handled when Registration will use messaging

please create a JIRA to track that

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/installations/InstallationRegistrationEndpoint.java [hunk 1]
```diff
@@ -73,7 +78,7 @@ public Response crossOriginForInstallations(@Context HttpHeaders headers) {
      *
      * <pre>
      * curl -u "variantID:secret"
-     *   -v -H "Accept: application/json" -H "Content-type: application/json"
+     *   -v -H "Accept: application/json" -H "Content-type: application/json" -H "aerogear-push-id: someid"
```
Comment (matzew) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/installations/InstallationRegistrationEndpoint.java]: let's have doc for that :-) what's it's for - when to inculde etc.

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/util/HttpRequestUtil.java [hunk 2]
```diff
@@ -86,6 +86,16 @@ public static String extractAeroGearSenderInformation(final HttpServletRequest r
     }
 
     /**
+     * Extract the header that identifies the push message.
+     *
+     * @param request
+     * @return
+     */
+    public static String extractPushIdentifier(HttpServletRequest request) {
```
Comment (matzew) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/util/HttpRequestUtil.java]: mind adding a test  ?

File: model/api/src/main/java/org/jboss/aerogear/unifiedpush/api/PushMessageInformation.java [hunk 3]
```diff
@@ -36,6 +36,10 @@
     private Date submitDate = new Date();
     private long totalReceivers;
 
+    private long appOpenCounter;
+    private Date firstOpenDate;
+    private Date lastOpenDate;
```
Comment (matzew) [inline model/api/src/main/java/org/jboss/aerogear/unifiedpush/api/PushMessageInformation.java]: we need to integrate this into the migration PR, that @qmx is currently working on for getting ready to be part of master

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/util/HttpRequestUtil.java [hunk 4]
```diff
@@ -86,6 +86,16 @@ public static String extractAeroGearSenderInformation(final HttpServletRequest r
     }
 
     /**
+     * Extract the header that identifies the push message.
+     *
+     * @param request
+     * @return
```
Comment (matzew) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/util/HttpRequestUtil.java]: please do full JavaDoc - on Java8 incomplete JavaDoc is crying like a baby. see also https://github.com/aerogear/aerogear-unifiedpush-server/pull/538

File: model/api/src/main/java/org/jboss/aerogear/unifiedpush/api/VariantMetricInformation.java [hunk 5]
```diff
@@ -97,4 +98,20 @@ public PushMessageInformation getPushMessageInformation() {
     public void setPushMessageInformation(PushMessageInformation pushMessageInformation) {
         this.pushMessageInformation = pushMessageInformation;
     }
+
+    /**
+     * To track how many time this variant has been opened after a Push Notification
+     *
```
Comment (matzew) [inline model/api/src/main/java/org/jboss/aerogear/unifiedpush/api/VariantMetricInformation.java]: return desc. missing

File: model/api/src/main/java/org/jboss/aerogear/unifiedpush/dao/VariantMetricInformationDao.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+package org.jboss.aerogear.unifiedpush.dao;
+import org.jboss.aerogear.unifiedpush.api.VariantMetricInformation;
+
+
+public interface VariantMetricInformationDao extends GenericBaseDao<VariantMetricInformation, String>{
+
+    /**
+     * Finds a VariantMetricInformation instance by the related variantID
+     *
+     * @param variantID that is associated with the VariantMetricInformation instance
```
Comment (matzew) [inline model/api/src/main/java/org/jboss/aerogear/unifiedpush/dao/VariantMetricInformationDao.java]: second param missing

File: model/api/src/main/java/org/jboss/aerogear/unifiedpush/dao/VariantMetricInformationDao.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.aerogear.unifiedpush.dao;
+import org.jboss.aerogear.unifiedpush.api.VariantMetricInformation;
+
+
```
Comment (matzew) [inline model/api/src/main/java/org/jboss/aerogear/unifiedpush/dao/VariantMetricInformationDao.java]: what's this class for?

File: model/jpa/testData/create_db_content.ddl [hunk 8]
```diff
@@ -74,6 +74,7 @@ CREATE TABLE PushMessageInformation (
   rawJsonMessage    VARCHAR(4500) DEFAULT NULL,
   submitDate        DATE          DEFAULT NULL,
   totalReceivers    BIGINT       NOT NULL,
+  appOpenCounter    BIGINT      DEFAULT 0,
```
Comment (matzew) [inline model/jpa/testData/create_db_content.ddl]: what about first/last open date?

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/APNsPushNotificationSender.java [hunk 9]
```diff
@@ -44,7 +44,7 @@
 import static org.jboss.aerogear.unifiedpush.message.util.ConfigurationUtils.tryGetIntegerProperty;
 
 @SenderType(iOSVariant.class)
-public class APNsPushNotificationSender implements PushNotificationSender {
+public class APNsPushNotificationSender  implements PushNotificationSender {
```
Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/APNsPushNotificationSender.java]: please dont commit reformatting

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/APNsPushNotificationSender.java [hunk 10]
```diff
@@ -65,12 +65,11 @@
      * Sends APNs notifications ({@link UnifiedPushMessage}) to all devices, that are represented by
      * the {@link Collection} of tokens for the given {@link iOSVariant}.
      */
-    public void sendPushMessage(final Variant variant, final Collection<String> tokens, final UnifiedPushMessage pushMessage, final NotificationSenderCallback callback) {
+    public void sendPushMessage(final Variant variant, final Collection<String> tokens, final UnifiedPushMessage pushMessage, final String pushMessageInformationId, final NotificationSenderCallback callback) {
```
Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/APNsPushNotificationSender.java]: looks like javadoc needs an update here

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/PushNotificationSender.java [hunk 11]
```diff
@@ -36,5 +38,5 @@
      * @param senderCallback invoked after submitting the request to the underlying push network to indicate the status
      *                       of the request (<code>success</code> or <code>error</code>
      */
-    void sendPushMessage(Variant variant, Collection<String> clientIdentifiers, UnifiedPushMessage pushMessage, NotificationSenderCallback senderCallback);
+    void sendPushMessage(Variant variant, Collection<String> clientIdentifiers, UnifiedPushMessage pushMessage, String pushMessageInformationId, NotificationSenderCallback senderCallback);
```
Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/PushNotificationSender.java]: javadoc update needed :)

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/PushNotificationSender.java [hunk 12]
```diff
@@ -27,6 +27,8 @@
  */
 public interface PushNotificationSender {
 
+    String AEROGEAR_PUSH_ID = "aerogear-push-id";
```
Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/PushNotificationSender.java]: should we have it here? hrm, not sure

File: service/src/test/java/org/jboss/aerogear/unifiedpush/service/AbstractBaseServiceTest.java [hunk 13]
```diff
@@ -21,17 +21,14 @@
 import org.apache.openejb.mockito.MockitoInjector;
 import org.apache.openejb.testing.MockInjector;
 import org.apache.openejb.testing.Module;
-import org.jboss.aerogear.unifiedpush.jpa.dao.impl.JPACategoryDao;
-import org.jboss.aerogear.unifiedpush.jpa.dao.impl.JPAInstallationDao;
-import org.jboss.aerogear.unifiedpush.jpa.dao.impl.JPAPushApplicationDao;
-import org.jboss.aerogear.unifiedpush.jpa.dao.impl.JPAPushMessageInformationDao;
-import org.jboss.aerogear.unifiedpush.jpa.dao.impl.JPAVariantDao;
+import org.jboss.aerogear.unifiedpush.jpa.dao.impl.*;
```
Comment (matzew) [inline service/src/test/java/org/jboss/aerogear/unifiedpush/service/AbstractBaseServiceTest.java]: please no wildcards

File: model/api/src/main/java/org/jboss/aerogear/unifiedpush/api/PushMessageInformation.java (hunk 3 above)
Comment (matzew) [inline model/api/src/main/java/org/jboss/aerogear/unifiedpush/api/PushMessageInformation.java]: Do we have test for these new properties ?

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/installations/InstallationRegistrationEndpoint.java (hunk 1 above)
Comment (sebastienblanc) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/installations/InstallationRegistrationEndpoint.java]: Will be added in our Registration doc https://aerogear.org/docs/unifiedpush/client-registration/ , seems that the good place ?

File: model/api/src/main/java/org/jboss/aerogear/unifiedpush/dao/VariantMetricInformationDao.java (hunk 7 above)
Comment (sebastienblanc) [inline model/api/src/main/java/org/jboss/aerogear/unifiedpush/dao/VariantMetricInformationDao.java]: To retrieve a specific VariantMetricInformation instance, we had no dao for this entity before.

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/PushNotificationSender.java (hunk 12 above)
Comment (sebastienblanc) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/PushNotificationSender.java]: Where should that be ? I thought it was a good place

File: model/api/src/main/java/org/jboss/aerogear/unifiedpush/dao/VariantMetricInformationDao.java (hunk 7 above)
Comment (matzew) [inline model/api/src/main/java/org/jboss/aerogear/unifiedpush/dao/VariantMetricInformationDao.java]: javadoc?

File: jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/installations/InstallationRegistrationEndpoint.java (hunk 1 above)
Comment (matzew) [inline jaxrs/src/main/java/org/jboss/aerogear/unifiedpush/rest/registry/installations/InstallationRegistrationEndpoint.java]: yeah - but perhaps here too, especually we generate the REST docs from here

File: model/api/src/main/java/org/jboss/aerogear/unifiedpush/dao/VariantMetricInformationDao.java (hunk 7 above)
Comment (sebastienblanc) [inline model/api/src/main/java/org/jboss/aerogear/unifiedpush/dao/VariantMetricInformationDao.java]: None of our DAO interfaces have Class level doc but yeah , I could add it for this one if needed.

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/PushNotificationSender.java (hunk 12 above)
Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/PushNotificationSender.java]: not sure, this is a header name, while the interface is the boundsry for
dispatching

we can keep it here, but nit sure the location is that good

On Wednesday, April 29, 2015, Sebastien Blanc <notifications@github.com>
wrote:

> In
> push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/PushNotificationSender.java
> <https://github.com/aerogear/aerogear-unifiedpush-server/pull/540#discussion_r29349251>
> :
>
> > @@ -27,6 +27,8 @@
> >   */
> >  public interface PushNotificationSender {
> >
> > +    String AEROGEAR_PUSH_ID = "aerogear-push-id";
>
> Where should that be ? I thought it was a good place
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/aerogear/aerogear-unifiedpush-server/pull/540/files#r29349251>
> .
>


-- 
Sent from Gmail Mobile

Comment (sebastienblanc) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/PushNotificationSender.java]: I understand but it's not only the header name, it's also a key of the payload.

Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/PushNotificationSender.java]: add it on the message class?

On Wednesday, April 29, 2015, Sebastien Blanc <notifications@github.com>
wrote:

> In
> push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/PushNotificationSender.java
> <https://github.com/aerogear/aerogear-unifiedpush-server/pull/540#discussion_r29352246>
> :
>
> > @@ -27,6 +27,8 @@
> >   */
> >  public interface PushNotificationSender {
> >
> > +    String AEROGEAR_PUSH_ID = "aerogear-push-id";
>
> I understand but it's not only the header name, it's also a key of the
> payload.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/aerogear/aerogear-unifiedpush-server/pull/540/files#r29352246>
> .
>


-- 
Sent from Gmail Mobile

Comment (sebastienblanc) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/PushNotificationSender.java]: as discussed on IRC we keep as it for now and we will see later if we can split the UnifiedMessage to hold "internal" information (https://issues.jboss.org/browse/AGPUSH-1381)

File: admin-ui/app/views/notification.html [hunk 14]
```diff
@@ -56,6 +56,18 @@
...
+                      <td colspan="2">{{ metric.appOpenCounter }}</td>
+                    </tr>
+                    <tr>
+                      <td>First time being opened:</td>
+                      <td colspan="2">{{ metric.firstOpenDate | date : 'd MMM, HH:mm:ss, yyyy'}}</td>
+                    </tr>
+                    <tr>
+                      <td>last time being opened:</td>
+                      <td colspan="2">{{ metric.lastOpenDate | date : 'd MMM, HH:mm:ss, yyyy'}}</td>
+                    </tr>
```
Comment (lfryc) [inline admin-ui/app/views/notification.html]: This file is no longer used in UI and needs to be removed in Beta1

**Comment by lfryc** [comment]
:+1: code-wise looks good

so far there is a way how to test this, but we will get there :-)

**Comment by sebastienblanc** [comment]
@lfryc I have a hacked "helloworld" Cordova that will read the aerogear-push-id from the payload and include it in the header when registration, this way you can test it https://github.com/sebastienblanc/unified-push-helloworld/tree/analytics . Also @edewit has a windows https://github.com/aerogear/aerogear-windows-push/pull/7 (that needs to be updated)

**Comment by edewit** [comment]
@sebastienblanc updated windows sdk

**Comment by corinnekrych** [comment]
@sebastienblanc @edewit I've done https://github.com/aerogear/aerogear-ios-push/pull/55
I still miss the background to foreground case as discussed in mailing list thread:
http://aerogear-dev.1069024.n5.nabble.com/aerogear-dev-Advanced-Analtyics-new-PR-and-call-to-the-client-tech-leads-td11546.html

**Comment by lfryc** [comment]
This PR exposes analytics to the new UI: https://github.com/sebastienblanc/aerogear-unified-push-server/pull/7

**Comment by matzew** [comment]
see https://github.com/aerogear/aerogear-unifiedpush-server/pull/543

**Pull request closed without merging**